packages/esm-form-entry-app/src/app/offline/caching.ts

```diff
@@ -18,3 +29,83 @@ export function setupOfflineDataSourcePrecaching() {
     );
   });
 }
+
+export function setupDynamicOfflineFormDataHandler() {
+  setupDynamicOfflineDataHandler({
+    id: 'esm-form-entry-app:form',
+    type: 'form',
+    displayName: 'Form entry',
+    async isSynced(identifier) {
+      const expectedUrls = await getCacheableFormUrls(identifier);
+      const absoluteExpectedUrls = expectedUrls.map((url) => window.origin + makeUrl(url));
+      const cache = await caches.open('omrs-spa-cache-v1');
+      const keys = (await cache.keys()).map((key) => key.url);
+      return absoluteExpectedUrls.every((url) => keys.includes(url));
+    },
+    async sync(identifier) {
+      const urlsToCache = await getCacheableFormUrls(identifier);
+      const cacheResults = await Promise.allSettled(
+        urlsToCache.map(async (urlToCache) => {
+          await messageOmrsServiceWorker({
+            type: 'registerDynamicRoute',
+            pattern: escapeRegExp(urlToCache),
+            strategy: 'network-first',
+          });
+
+          await openmrsFetch(urlToCache, {
+            headers: {
+              [omrsOfflineCachingStrategyHttpHeaderName]: 'network-first',
+            },
+          });
+        }),
+      );
+
+      if (cacheResults.some((x) => x.status === 'rejected')) {
+        throw new Error(`Some form data could not be properly downloaded. (Form UUID: ${identifier})`);
+      }
+    },
+  });
+}
+
+async function getCacheableFormUrls(formUuid: string) {
+  const getFormRes = await openmrsFetch<FormEncounter>(`/ws/rest/v1/form/${formUuid}?v=full`);
+  const form = getFormRes.data;
+  const getClobdataRes = await openmrsFetch(`/ws/rest/v1/clobdata/${form.resources[0].valueReference}?v=full`);
+  const clobdata = getClobdataRes.data;
+
+  if (!form || !clobdata) {
+    throw new Error(`The form data could not be loaded from the server. (Form UUID: ${formUuid})`);
+  }
+
+  const formSchemaCompiler = new FormSchemaCompiler();
+  const formSchema = formSchemaCompiler.compileFormSchema(
+    {
+      ...form,
+      pages: clobdata.pages ?? [],
+      referencedForms: clobdata.referencedForms ?? [],
+      processor: clobdata.processor,
+    },
+    [],
+  ) as FormSchema;
+
+  const conceptLang = (window as any).i18next?.language?.substring(0, 2).toLowerCase() || 'en';
```

Comment:
I am not sure if caching only the current language is sufficient. I think there are pros and cons.

Pro:
* Less network requests.
* Less data in the cache (potentially helps with storage quotas).
* I am assuming that in 99.99% of the cases users will only ever use one language. I assume that this language will also be selected _before_ downloading the form. It should probably work out fine most of the time.

Con:
Really only one flow that can fail: The user downloads a form, changes the language and then immediately goes offline without reloading the page. The concept labels would then not be precached.

I guess the potential remedies would either be:
* To listen to a language change and then redownload the concepts with the right language. I think that makes the most sense, but I am currently not sure how to ideally listen to a language change. Open for any help here!
* To download concepts in all (known) languages. That can obviously make the cache explode. Also, I don't know if we have a list of all known languages.

One other thing that is related to the language, but not to offline specifically: Does anybody know what happens (in online mode) when a concept in a specific language does not exist in the backend? Does the network request fail? If so, that would probably lead to the form not displaying anything in the current version.

Reply:
> Does anybody know what happens (in online mode) when a concept in a specific language does not exist in the backend?

We return the concept in the default language for the system.

> To download concepts in all (known) languages. That can obviously make the cache explode. Also, I don't know if we have a list of all known languages.

We do have a list of supported languages; however, as a practical matter, users will generally at most switch their language once to their preferred language and probably never look at the toggle again. If the issue comes up in demos and will go to bat to say that switching languages is generally not something users do, so while this is a potential problem, I don't think its one worth solving.

(The main use-case I can think of where there might be a valid need to do something like this is if we explicitly allowed multiple user to use the same device offline, but since this is not something we allow, it's not worth the complexity overhead to implement, IMO).

Reply:
> We return the concept in the default language for the system.
That's great, thanks!

> [...] so while this is a potential problem, I don't think its one worth solving.
Awesome, that was also my thought. If bad comes to worse, we could potentially also implement a fallback approach while offline (though that might become a bit finnicky on the code side).